server/src/main/java/org/apache/iotdb/db/utils/TimeValuePairUtils.java

```diff
@@ -107,4 +115,228 @@ public static TimeValuePair getEmptyTimeValuePair(TSDataType dataType) {
+        long thatUB = that.getUpperBound(thatIndex);
+        if (thisLB <= thatLB) {
+          result.mergeLast(thisLB, thisUB);
+          thisIndex++;
+        } else {
+          result.mergeLast(thatLB, thatUB);
+          thatIndex++;
+        }
+      }
+      // merge the remaining intervals
+      Intervals remainingIntervals = thisIndex < thisSize ? this : that;
+      int remainingIndex = thisIndex < thisSize ? thisIndex : thatIndex;
+      mergeRemainingIntervals(remainingIndex, remainingIntervals, result);
+
+      return result;
+    }
+
+    private void mergeRemainingIntervals(
+        int remainingIndex, Intervals remainingIntervals, Intervals result) {
+      for (int i = remainingIndex; i < remainingIntervals.getIntervalSize(); i++) {
+        long lb = remainingIntervals.getLowerBound(i);
+        long ub = remainingIntervals.getUpperBound(i);
+        result.mergeLast(lb, ub);
+      }
+    }
+
+    /**
+     * Merge an interval of [lowerBound, upperBound] with the last interval if they can be merged,
+     * or just add it as the last interval if its lowerBound is larger than the upperBound of the
+     * last interval. If the upperBound of the new interval is less than the lowerBound of the last
+     * interval, nothing will be done.
+     *
+     * @param lowerBound
+     * @param upperBound
+     */
+    private void mergeLast(long lowerBound, long upperBound) {
+      if (getIntervalSize() == 0) {
+        addInterval(lowerBound, upperBound);
+        return;
+      }
+      int lastIndex = getIntervalSize() - 1;
+      long lastLB = getLowerBound(lastIndex);
+      long lastUB = getUpperBound(lastIndex);
+      if (lowerBound > lastUB + 1) {
+        // e.g., last [3, 5], new [7, 10], just add the new interval
+        addInterval(lowerBound, upperBound);
+        return;
+      }
+      if (upperBound < lastLB - 1) {
+        // e.g., last [7, 10], new [3, 5], do nothing
+        return;
+      }
+      // merge the new interval into the last one
+      setLowerBound(lastIndex, Math.min(lastLB, lowerBound));
+      setUpperBound(lastIndex, Math.max(lastUB, upperBound));
+    }
+
+    public Intervals not() {
+      if (isEmpty()) {
+        return ALL_INTERVAL;
+      }
+      Intervals result = new Intervals();
+      long firstLB = getLowerBound(0);
+      if (firstLB != Long.MIN_VALUE) {
+        result.addInterval(Long.MIN_VALUE, firstLB - 1);
+      }
+
+      int intervalSize = getIntervalSize();
+      for (int i = 0; i < intervalSize - 1; i++) {
+        long currentUB = getUpperBound(i);
+        long nextLB = getLowerBound(i + 1);
+        if (currentUB + 1 <= nextLB - 1) {
+          result.addInterval(currentUB + 1, nextLB - 1);
+        }
+      }
+
+      long lastUB = getUpperBound(result.getIntervalSize() - 1);
+      if (lastUB != Long.MAX_VALUE) {
+        result.addInterval(lastUB + 1, Long.MAX_VALUE);
+      }
+      return result;
+    }
+  }
```

Comment:
Is those code duplicated with the org.apache.iotdb.cluster.utils.PartitionUtils.Intervals ? if yes, why repeat coding in two places？
Don't repeat yourself.

Reply:
I'll make some improvements